### ISSUE TITLE
is there a way to identify the tenant by using Header instead of sub-domain or sub folder.

### ISSUE THREAD

**Issue by additionsolutions** [django-tenants/django-tenants#658]
### Discussed in https://github.com/django-tenants/django-tenants/discussions/611

<div type='discussions-op-text'>

<sup>Originally posted by **praburocking** June  6, 2021</sup>
Hi, is there a way to identify the tenant using the request Header like X-USER-ID=123245, instead of using sub-domain or sub folder?</div>

**Comment by additionsolutions** [comment]
This is what I replied in discussion. Let me know if this approach is correct then I can do a pull request and close the issue.

######################################################################
Yes, Its possible with little hack. Very similar to what is done in 'django_tenant_schemas'

In my case I was passing 'schema_name' in request header.
Here is how I did it.

You need request object in middleware to check value passed in request header.
Hence we need to pass request for get_tenant() method.
Modify main.py in TenantMainMiddleware as follows:

    def process_request(self, request):
       ....
       try:
                tenant = self.get_tenant(domain_model, hostname, request)
       ....
Note 'request' added as additional parameter, this will help us when we override get_tenant in our custom middleware

Add your custom middleware

 class XHeaderTenantMiddleware(TenantMainMiddleware):
     """
     Determines tenant by the value of the ``X-???`` in HTTP header.
     """
     def get_tenant(self, model, hostname, request):
        .....
        < Write your logic to get tenant based on value you passed in header using request.META.get() >
        .....
 
         domain = model.objects.select_related('tenant').get(tenant__schema_name=schema_name)
         return domain.tenant
In this example schema_name is used to get tenant.

Do not forgot to mention your custom middleware in settings.py

Hope this helps.

Caveat: I have used this logic in production using 'django_tenant_schemas' but not with 'django_tenants', so its not fully tested.

**Comment by ranchoteca** [comment]
Im interested in this topic.

Could be nice this approach for certain projects.

**Comment by praburocking** [comment]
yes, if we provide an option to handle tenant switching using custom middleware, it will be helpful. I think this option is already implemented in django_tenent_schemas.

**Comment by tomturner** [comment]
You would have to write your own middleware to do this. Good luck :)

**Comment by bluebeel** [comment]
Yep you can do it. We did it like it:
```python
# models.py
from django.db import models
from django_tenants.models import TenantMixin, DomainMixin
import uuid


class Tenant(TenantMixin):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, null=False, blank=False)
    name = models.CharField(max_length=100, null=False, blank=False)
    url = models.URLField(blank=True, null=True)
    created_on = models.DateField(auto_now_add=True)

    # default true, schema will be automatically created and synced when it is saved
    auto_create_schema = True

    def __str__(self):
        return f"{self.schema_name} - {self.id} - {self.url}"


class Domain(DomainMixin):
    pass
```
The middleware can probably be simplified
```python
# middleware.py
from shayp.web.apps.tenants.models import Tenant
from typing import Type
from django.db import connection
from django.http.request import HttpRequest

from django_tenants.utils import get_tenant_model
from django_tenants.middleware import TenantMainMiddleware


class XTenantMiddleware(TenantMainMiddleware):
    def process_request(self, request: HttpRequest):
        # Connection needs first to be at the public schema, as this is where
        # the tenant metadata is stored.

        connection.set_schema_to_public()
        hostname = self.hostname_from_request(request)

        tenant_model = get_tenant_model()
        try:
            tenant = self.get_tenant(tenant_model, request)
        except tenant_model.DoesNotExist:
            self.no_tenant_found(request, hostname)
            return

        tenant.domain_url = hostname
        request.tenant = tenant
        connection.set_tenant(request.tenant)
        self.setup_url_routing(request)

    def get_tenant(self, model: Type[Tenant], request: HttpRequest) -> Type[Tenant]:
        tenant_id = request.META.get("HTTP_X_TENANT")
        # if the header is missing in our case we consider it's the default tenant he wants to access
        if not tenant_id:
            return model.objects.get(schema_name="default_tenant")

        return model.objects.get(id=tenant_id)
```

Do not forget to put the middleware in the settings :)

**Comment by tomturner** [comment]
Consolidating the backlog: this is the same underlying request as #1206, which is where identifying the tenant from a request header is being tracked. Closing here so the discussion stays in one place — please add anything from your case over there, and thanks for reporting it.
